### PR TITLE
[mirrativ] add support for mirrativ.com

### PR DESCRIPTION
**Pull request by nao20010128nao**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

Note that the code is taken from [my fork](https://github.com/ytdl-patched/ytdl-patched/blob/ytdlp/yt_dlp/extractor/mirrativ.py), which is my own work.

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This PR adds support for mirrativ.com, which is a streaming service in Japan.

example URL for live: https://www.mirrativ.com/live/xvbjsvT96UFjkPqaRj9ajg (running as of writing, but you can find new one at https://mirrativ.com/ )
example URL for user: https://www.mirrativ.com/user/7077880

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -88,7 +88,7 @@ The major new features from the latest release of [blackjack4494/yt-dlc](https:/
 
 * **Aria2c with HLS/DASH**: You can use `aria2c` as the external downloader for DASH(mpd) and HLS(m3u8) formats
 
-* **New extractors**: AnimeLab, Philo MSO, Spectrum MSO, SlingTV MSO, Rcs, Gedi, bitwave.tv, mildom, audius, zee5, mtv.it, wimtv, pluto.tv, niconico users, discoveryplus.in, mediathek, NFHSNetwork, nebula, ukcolumn, whowatch, MxplayerShow, parlview (au), YoutubeWebArchive, fancode, Saitosan, ShemarooMe, telemundo, VootSeries, SonyLIVSeries, HotstarSeries, VidioPremier, VidioLive, RCTIPlus, TBS Live, douyin, pornflip, ParamountPlusSeries, ScienceChannel, Utreon, OpenRec
+* **New extractors**: AnimeLab, Philo MSO, Spectrum MSO, SlingTV MSO, Rcs, Gedi, bitwave.tv, mildom, audius, zee5, mtv.it, wimtv, pluto.tv, niconico users, discoveryplus.in, mediathek, NFHSNetwork, nebula, ukcolumn, whowatch, MxplayerShow, parlview (au), YoutubeWebArchive, fancode, Saitosan, ShemarooMe, telemundo, VootSeries, SonyLIVSeries, HotstarSeries, VidioPremier, VidioLive, RCTIPlus, TBS Live, douyin, pornflip, ParamountPlusSeries, ScienceChannel, Utreon, OpenRec, Mirrativ
```
Comment (pukkandan) [inline README.md:91]:
```suggestion
* **New extractors**: AnimeLab, Philo MSO, Spectrum MSO, SlingTV MSO, Rcs, Gedi, bitwave.tv, mildom, audius, zee5, mtv.it, wimtv, pluto.tv, niconico users, discoveryplus.in, mediathek, NFHSNetwork, nebula, ukcolumn, whowatch, MxplayerShow, parlview (au), YoutubeWebArchive, fancode, Saitosan, ShemarooMe, telemundo, VootSeries, SonyLIVSeries, HotstarSeries, VidioPremier, VidioLive, RCTIPlus, TBS Live, douyin, pornflip, ParamountPlusSeries, ScienceChannel, Utreon, OpenRec
```

I will add these at release. If everyone starts doing this, there would be unnecessary conflicts

File: yt_dlp/extractor/mirrativ.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        'only_matching': True,
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage('https://www.mirrativ.com/live/%s' % video_id, video_id)
+        live_response = self._download_json(self.LIVE_API_URL % video_id, video_id)
+        self.assert_error(live_response)
+
+        hls_url = live_response.get('archive_url_hls') or live_response.get('streaming_url_hls')
```
Comment (pukkandan) [inline yt_dlp/extractor/mirrativ.py:33]:
```suggestion
        hls_url = traverse_obj(live_response, 'archive_url_hls', 'streaming_url_hls')
```

or 
```py
        hls_url = dict_get(live_response, ('archive_url_hls', 'streaming_url_hls'))
```

File: yt_dlp/extractor/mirrativ.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        is_live = bool(live_response.get('is_live'))
+        was_live = bool(live_response.get('is_archive'))
+        if not hls_url:
+            raise ExtractorError('Neither archive nor live is available.', expected=True)
+
+        formats = self._extract_m3u8_formats(
+            hls_url, video_id,
+            ext='mp4', entry_protocol='m3u8',
```
Comment (pukkandan) [inline yt_dlp/extractor/mirrativ.py:41]: Any reason why this is being forced to use ffmpeg?

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/mirrativ.py (hunk 3 above)
Comment (nao20010128nao) [reply yt_dlp/extractor/mirrativ.py:41]: this is either live stream or archive    
latter don't need to be ffmpeg. will fix

File: yt_dlp/extractor/mirrativ.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            for live in lives:
+                # !(live.is_archive || live.is_live) === !live.is_archive && !live.is_live
+                if not live.get('is_archive') and not live.get('is_live'):
+                    # neither archive nor live is available, so skip it
+                    # or the service will ban your IP address for a while
+                    continue
+                live_id = live.get('live_id')
+                url = 'https://www.mirrativ.com/live/%s' % live_id
+                entries.append(self.url_result(url, video_id=live_id, video_title=live.get('title')))
+            page = api_response.get('next_page')
```
[the remark this answers is not in the document]
Comment (nao20010128nao) [reply yt_dlp/extractor/mirrativ.py:131]: okay, will fix

File: yt_dlp/extractor/mirrativ.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+            } | {k: traverse_obj(formats, (0, k)) for k in keys_to_copy})
+        self._sort_formats(formats)
+
+        title = self._og_search_title(webpage, default=None) or self._search_regex(
+            r'<title>\s*(.+?) - Mirrativ\s*</title>', webpage) or live_response.get('title')
+        description = live_response.get('description')
+        thumbnail = live_response.get('image_url')
+
+        if live_response.get('started_at') and live_response.get('ended_at'):
+            duration = live_response.get('ended_at') - live_response.get('started_at')
```
[the remark this answers is not in the document]
Comment (nao20010128nao) [reply yt_dlp/extractor/mirrativ.py:60]: `int_or_none` doesn't make a sense if one of `started_at` or `ended_at` aren't sense. do you mean `try_get`?

[the remark this answers is not in the document]
Comment (nao20010128nao) [reply yt_dlp/extractor/mirrativ.py]:
```
duration = try_get(live_response, lambda x: x['ended_at'] - x['started_at'])
```

[the remark this answers is not in the document]
Comment (pukkandan) [reply yt_dlp/extractor/mirrativ.py]: I was thinking of passing the stuff in the if through int_or_none, but this is even better

File: yt_dlp/extractor/mirrativ.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+            m3u8_id='hls', live=is_live)
+        rtmp_url = live_response.get('streaming_url_edge')
+        if rtmp_url:
+            keys_to_copy = ('width', 'height', 'vcodec', 'acodec', 'tbr')
+            formats.append({
+                'format_id': 'rtmp',
+                'url': rtmp_url,
+                'protocol': 'rtmp',
+                'ext': 'mp4',
+            } | {k: traverse_obj(formats, (0, k)) for k in keys_to_copy})
```
Comment (pukkandan) [inline yt_dlp/extractor/mirrativ.py:53]: `|` is only supported in py9

File: yt_dlp/extractor/mirrativ.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        while page is not None:
+            api_response = self._download_json(
+                self.LIVE_HISTORY_API_URL % (user_id, page), user_id,
+                note='Downloading page %d' % page)
+            self.assert_error(api_response)
+            lives = api_response.get('lives')
+            if not lives:
+                break
+            for live in lives:
+                # !(live.is_archive || live.is_live) === !live.is_archive && !live.is_live
```
Comment (pukkandan) [inline yt_dlp/extractor/mirrativ.py:112]:
```suggestion
```

**Pull request merged**